base/special/gamma.jl

```diff
@@ -277,29 +280,34 @@ function zeta(s::Union{Int,Float64,Complex{Float64}},
 
     isnan(x) && return oftype(ζ, imag(z)==0 && isa(s,Int) ? x : Complex(x,x))
 
-    cutoff = 7 + real(m) + imag(m) # TODO: this cutoff is too conservative?
+    cutoff = 7 + real(m) + abs(imag(m)) # TODO: this cutoff is too conservative?
     if x < cutoff
         # shift using recurrence formula
         xf = floor(x)
-        if x <= 0 && xf == z
-            if isa(s, Int)
-                iseven(s) && return oftype(ζ, Inf)
-                x == 0 && return oftype(ζ, inv(x))
-            end
-            throw(DomainError()) # or return NaN?
-        end
         nx = Int(xf)
         n = ceil(Int,cutoff - nx)
-        ζ += inv_oftype(ζ, z)^s
-        for ν = -nx:-1:1
-            ζₒ= ζ
-            ζ += inv_oftype(ζ, z + ν)^s
-            ζ == ζₒ && break # prevent long loop for large -x > 0
-                             # FIXME: still slow for small m, large Im(z)
+        minus_s = -s
+        if nx < 0 # x < 0
+            # need to use (-z)^(-s) recurrence to be correct for real z < 0
+            # [the general form of the recurrence term is (z^2)^(-s/2)]
+            minus_z = -z
+            ζ += pow_oftype(ζ, minus_z, minus_s) # ν = 0 term
+            if xf != z
+                ζ += pow_oftype(ζ, z - nx, minus_s) # real(z - nx) > 0, so use correct branch cut
+                # otherwise, if xf==z, then the definition skips this term
+            end
+            for ν in (real(s) > 0 ? (-nx-1:-1:1) : (1:-nx-1))
+                ζₒ= ζ
+                ζ += pow_oftype(ζ, minus_z - ν, minus_s)
+                ζ == ζₒ && break # prevent long loop for large -x > 0
+                                 # FIXME: still slow for small m, large Im(z)
+            end
+        else # x ≥ 0 && z != 0
+            ζ += pow_oftype(ζ, z, minus_s)
         end
-        for ν = max(1,1-nx):n-1
+        for ν in (real(s) > 0 ? (max(1,1-nx):n-1) : (n-1:-1:max(1,1-nx)))
```

Comment:
Looks like it kills type inference, especially because I have two different iterator types.  #15946.   I will update the patch to duplicate the loops.